Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Spawnable/PrefabDocument.cpp

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <AzToolsFramework/Prefab/PrefabDomUtils.h>
+#include <AzToolsFramework/Prefab/Spawnable/PrefabDocument.h>
+#include <AzToolsFramework/Prefab/Spawnable/PrefabProcessorContext.h>
+
+namespace AzToolsFramework::Prefab::PrefabConversionUtils
+{
+    PrefabDocument::PrefabDocument(AZStd::string name)
+        : m_name(AZStd::move(name))
+        , m_instance(AZStd::make_unique<AzToolsFramework::Prefab::Instance>())
+    {
+        m_instance->SetTemplateSourcePath(AZ::IO::Path("InMemory") / name);
+    }
+
+    bool PrefabDocument::SetPrefabDom(const PrefabDom& prefab)
+    {
+        if (ConstructInstanceFromPrefabDom(prefab))
+        {
+            constexpr bool copyConstStrings = true;
+            m_dom.CopyFrom(prefab, m_dom.GetAllocator(), copyConstStrings);
+            return true;
+        }
+        else
+        {
+            return false;
+        }
+    }
+
+    bool PrefabDocument::SetPrefabDom(PrefabDom&& prefab)
+    {
+        if (ConstructInstanceFromPrefabDom(prefab))
+        {
+            m_dom = AZStd::move(prefab);
+            return true;
+        }
+        else
+        {
+            return false;
+        }
+    }
+
+    const AZStd::string& PrefabDocument::GetName() const
+    {
+        return m_name;
+    }
+
+    const PrefabDom& PrefabDocument::GetDom() const
+    {
+        if (m_isDirty)
+        {
+            m_isDirty = !PrefabDomUtils::StoreInstanceInPrefabDom(*m_instance, m_dom);
+        }
+        return m_dom;
+    }
+
+    PrefabDom&& PrefabDocument::TakeDom()
+    {
+        if (m_isDirty)
```

Comment:
Shouldn't `m_isDirty` always still be true after this call, because we've moved the DOM and subsequent get/take calls will have potentially invalid data cached in `m_dom`?

Reply:
You are correct, well spotted!
The idea is that the `Instance` and `PrefabDom` are presented to represent the same data, so if the DOM is moved out and leaves an empty DOM behind then the `Instance` should be reset to represent the `PrefabDom` that's left after the move. Fixing this now.